File: PermutationsII.py
# permutationII

class Solution:
    """
    @param nums: A list of Integers.
    @return: A list of permutations.
    """
    def __init__(self):
        self.visited = []

    def permuteUnique(self, S):
        result = []

        if S == None or not len(S):
            return result
        current = []

        for i in range(len(S)):
            self.visited.append(0)

        S.sort()
        self.Helper(result, current, self.visited, S)

        return result

    def Helper(self, result, current, visited, S):
        # debug
        # print "jump into Helper function"
        # print "#"*30
        # print visited

        if (len(current) == len(S)):
            result.append(current[:])

        for i in range(len(S)):
            if self.visited[i] == 1 or i != 0 and S[i] == S[i - 1] and self.visited[i - 1] == 0:
                continue
            # debug
            # print "assign value to self.visited"
            # print "i = %d"%i

            self.visited[i] = 1
            current.append(S[i])
            self.Helper(result, current, self.visited, S)
            current.pop()
            self.visited[i] = 0



#test
if __name__ == '__main__':
    S1 = []
    S2 = None
    S3 = [1,2,3]
    S4 = [1,2,2]
    a = Solution()
    # print "-"*20
    # print a.permuteUnique(S1)
    # print "-"*20
    # print a.permuteUnique(S2)
    print "-"*20
    print a.permuteUnique(S3)
    print "-"*20
    print a.permuteUnique(S4)

File: java edition/permutationsII.java
import java.util.*;

class permutationsII{
    public static void main(String[] args){
        ArrayList<Integer> test = new ArrayList<Integer>();
        test.add(1);
        test.add(2);
        test.add(2);
        ArrayList<ArrayList<Integer>> result = permuteUnique(test);
        for(int i = 0; i < result.size(); i++){
            System.out.println(result.get(i));
        }
    }

    public static ArrayList<ArrayList<Integer>> permuteUnique(ArrayList<Integer> nums){
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if(nums == null || nums.size() == 0){
            return result;
        }
        ArrayList<Integer> list = new ArrayList<Integer>();
        int[] visited = new int[nums.size()];

        Collections.sort(nums);
        helper(result, list, visited, nums);
        return result;
    }

    private static void helper(ArrayList<ArrayList<Integer>> result, ArrayList<Integer> list, int[] visited,
            ArrayList<Integer> nums){
        if(list.size() == nums.size()){
            result.add(new ArrayList<Integer>(list));
            return;
        }

        for(int i = 0; i < nums.size(); i++){
            if(visited[i] == 1 || (i > 0 && nums.get(i) == nums.get(i - 1) && visited[i - 1] == 0)){
                continue;
            }
            visited[i] = 1;
            list.add(nums.get(i));
            helper(result, list, visited, nums);
            list.remove(list.size() - 1);
            visited[i] = 0;
        }
    }
}
